### PR TITLE
Add `EventLogPosition` python wrapper

### PR DESCRIPTION
**Pull request by edgarriba**
implements #45 based on the c++ codebase

Introduces `EventLogPosition` wrapper to make search more flexible

### CODE REVIEW

File: py/farm_ng/core/events_file_reader.py [hunk 1]
```diff
@@ -60,9 +68,9 @@ def __exit__(self, type: Any, value: Any, traceback: Any) -> None:
 
     def __repr__(self) -> str:
         return (
-            f"file_name: {str(self.file_name)} "
-            + f"file_stream: {self._file_stream} "
-            + f"is_open: {self.is_open()} "
+            f"file_name: {str(self.file_name)}\n"
+            + f"file_stream: {self._file_stream}\n"
+            + f"is_open: {self.is_open()}"
```
Comment (Hackerman342) [inline py/farm_ng/core/events_file_reader.py:73]:
```suggestion
            "\n".join([
                f"file_name: {str(self.file_name)}",
                f"file_stream: {self._file_stream}",
                f"is_open: {self.is_open()}"
            ])
```

File: py/farm_ng/core/events_file_reader.py [hunk 2]
```diff
@@ -95,19 +107,28 @@ def close(self) -> bool:
...
+        self._event_index = []
 
-    def uris(self) -> List[Uri]:
-        if not self._event_index:
+    def get_uris(self) -> List[Uri]:
+        if len(self._event_index) == 0:
             self._build_event_index()
-        return [string_to_uri(key) for key in sorted(self._event_index.keys())]
+
+        uris_set: Set[EventLogPosition] = set()
```
Comment (Hackerman342) [inline py/farm_ng/core/events_file_reader.py:116]:
```suggestion
        uris_set: Set[str] = set()
```

File: py/farm_ng/core/events_file_reader.py [hunk 3]
```diff
@@ -95,19 +107,28 @@ def close(self) -> bool:
...
+        for event_log in self.event_index:
+            uris_set.add(uri_to_string(event_log.event.uri))
+
+        return [string_to_uri(x) for x in sorted(uris_set)]
```
Comment (Hackerman342) [inline py/farm_ng/core/events_file_reader.py:120]: Can you not make a set of uri's to begin with?

File: py/farm_ng/core/events_file_reader.py [hunk 4]
```diff
@@ -95,19 +107,28 @@ def close(self) -> bool:
...
 
     def has_uri(self, uri: Uri) -> bool:
-        if self._event_index is None:
+        if len(self._event_index) == 0:
             return False
```
Comment (Hackerman342) [inline py/farm_ng/core/events_file_reader.py:124]: If using one liner below
```suggestion
```

File: py/farm_ng/core/events_file_reader.py [hunk 5]
```diff
@@ -95,19 +107,28 @@ def close(self) -> bool:
...
-        return uri_to_string(uri) in self._event_index.keys()
 
-    def _read_next_event(self) -> Event:
+        for idx in self.event_index:
+            if uri_to_string(uri) == uri_to_string(idx.event.uri):
+                return True
+        return False
```
Comment (Hackerman342) [inline py/farm_ng/core/events_file_reader.py:129]: Optionally
```suggestion
        return uri_to_string(uri) in [uri_to_string(idx.event.uri) for idx in self.event_index]
```

Comment (Hackerman342) [reply py/farm_ng/core/events_file_reader.py:129]: Probably worse for performance if `self.event_index` is long

File: py/farm_ng/core/events_file_reader.py [hunk 6]
```diff
@@ -45,10 +55,8 @@ def __init__(self, file_name: Union[str, Path]) -> None:
         self._file_stream: Optional[IO] = None
         self._file_length: int = 0
 
-        # store the bytes offset in a dictionary where:
-        # - the key is the string representation of the Uri of the message
-        # - the value is a list with tuples of Event and its offsets to messages with that Uri
-        self._event_index: DefaultDict[str, List[Tuple[Event, int]]] = defaultdict(list)
+        # store index of events in a list to seek faster
+        self._event_index: List[EventLogPosition] = []
```
Comment (Hackerman342) [inline py/farm_ng/core/events_file_reader.py:59]: Clearer that it is all events, not the index of a single event
```suggestion
        self._events_index: List[EventLogPosition] = []
```

File: py/farm_ng/core/events_file_reader.py [hunk 7]
```diff
@@ -73,6 +81,10 @@ def file_length(self) -> int:
     def file_name(self) -> Path:
         return self._file_name
 
+    @property
+    def event_index(self) -> List[EventLogPosition]:
+        return self._event_index
```
Comment (Hackerman342) [inline py/farm_ng/core/events_file_reader.py:86]: Clearer that it is all events, not the index of a single event
```suggestion
    def events_index(self) -> List[EventLogPosition]:
        return self._events_index
```

File: py/farm_ng/core/events_file_reader.py [hunk 8]
```diff
@@ -135,49 +157,46 @@ def _build_event_index(self) -> None:
...
+        return len(self.get_events(uri))
 
-    def get_event(self, uri: Uri, frame_id: int) -> Tuple[Event, int]:
-        assert self.has_uri(uri)
-        assert frame_id < self.num_events(uri)
-        return self.events(uri)[frame_id]
+    def get_index(self) -> List[EventLogPosition]:
+        if len(self._event_index) == 0:
+            self._build_event_index()
+        return self._event_index
```
Comment (Hackerman342) [inline py/farm_ng/core/events_file_reader.py:181]:
```suggestion
    def get_events_index(self) -> List[EventLogPosition]:
        if len(self._events_index) == 0:
            self._build_events_index()
        return self._events_index
```

File: py/farm_ng/core/events_file_reader.py [hunk 9]
```diff
@@ -135,49 +157,46 @@ def _build_event_index(self) -> None:
...
         message: Message = message_cls()
         message.ParseFromString(payload)
         return message
 
     def read(self) -> Tuple[Event, Message]:
         assert self.is_open()
-        event = self._read_next_event()
-        return event, self.read_message(event)
+        event_log: EventLogPosition = self.read_next_event()
+        return event_log.event, self.read_message(event_log)
```
Comment (Hackerman342) [inline py/farm_ng/core/events_file_reader.py:199]: `event_log` could be confusing for a single event
```suggestion
        event: EventLogPosition = self.read_next_event()
        return event.event, self.read_message(event)
```

**Review by Hackerman342 — APPROVED** [review]

**Review by ethanrublee — COMMENTED** [review]
I would suggest simplifying the api by removing the uri filtering from this code. Uri filtering can easily be done while iterating over the event pod list, in an application. I think until there are clear common use cases it's better to have less functionality in this reader class so that there is less to maintain and it becomes stable faster. Recommend following c++ interface.

File: py/farm_ng/core/events_file_reader.py (hunk 3 above)
Comment (edgarriba) [reply py/farm_ng/core/events_file_reader.py:120]: just replicating the previous behavior where we based on a dict with unique uris, in the this new approach using EventLogPosition we build the index a list not dict

File: py/farm_ng/core/events_file_reader.py [hunk 10]
```diff
@@ -95,19 +113,19 @@ def close(self) -> bool:
...
+    def get_uris(self) -> List[Uri]:
+        if len(self._events_index) == 0:
             self._build_event_index()
-        return [string_to_uri(key) for key in sorted(self._event_index.keys())]
 
-    def has_uri(self, uri: Uri) -> bool:
-        if self._event_index is None:
-            return False
-        return uri_to_string(uri) in self._event_index.keys()
+        uris_set: Set[str] = set()
```
Comment (ethanrublee) [inline py/farm_ng/core/events_file_reader.py:122]: Ideally this is built once during the index constructor so that it's constant small time to call.

**Review by ethanrublee — APPROVED** [review]

File: py/tests/test_core.py [hunk 11]
```diff
@@ -61,10 +61,10 @@ def test_smoke(self, log_file: Path) -> None:
         with EventsFileReader(log_file) as reader:
             assert reader.is_open()
             assert len(reader.events_index) == 0
-            uris = reader.get_uris()
+            uris = sorted([*{x.event.uri.path for x in reader.get_index()}])
```
Comment (ethanrublee) [inline py/tests/test_core.py:64]: nice one liner!

Comment (edgarriba) [reply py/tests/test_core.py:64]: very ugly actually :smile:

**Pull request merged**